src/zino/tasks/bgpstatemonitortask.py

```diff
@@ -0,0 +1,393 @@
+import logging
+from dataclasses import dataclass
+from ipaddress import ip_address
+from typing import Any, Iterable, Optional
+
+from pyasn1.type.univ import OctetString
+
+from zino.snmp import SparseWalkResponse
+from zino.statemodels import BgpAdminStatus, BGPEvent, BgpOperState, BgpStyle, IPAddress
+from zino.tasks.task import Task
+
+_logger = logging.getLogger(__name__)
+
+TIME_BEFORE_OPER_DOWN_ALERT = 600
+
+JUNIPER_TRANSLATION_MAP = [
+    ("peer_state", "jnxBgpM2PeerState"),
+    ("peer_admin_status", "jnxBgpM2PeerStatus"),
+    ("peer_remote_address", "jnxBgpM2PeerRemoteAddr"),
+    ("peer_remote_as", "jnxBgpM2PeerRemoteAs"),
+    ("peer_fsm_established_time", "jnxBgpM2PeerFsmEstablishedTime"),
+]
+CISCO_TRANSLATION_MAP = [
+    ("peer_state", "cbgpPeer2State"),
+    ("peer_admin_status", "cbgpPeer2AdminStatus"),
+    ("peer_remote_address", "cbgpPeer2RemoteAddr"),
+    ("peer_remote_as", "cbgpPeer2RemoteAs"),
+    ("peer_fsm_established_time", "cbgpPeer2FsmEstablishedTime"),
+]
+GENERAL_TRANSLATION_MAP = [
+    ("peer_state", "bgpPeerState"),
+    ("peer_admin_status", "bgpPeerAdminStatus"),
+    ("peer_remote_address", "bgpPeerRemoteAddr"),
+    ("peer_remote_as", "bgpPeerRemoteAs"),
+    ("peer_fsm_established_time", "bgpPeerFsmEstablishedTime"),
+]
+LOCAL_AS_OBJECTS = {
+    "juniper": ("BGP4-V2-MIB-JUNIPER", "jnxBgpM2PeerLocalAs"),
+    "cisco": ("CISCO-BGP4-MIB", "cbgpLocalAs"),
+    "general": ("BGP4-MIB", "bgpLocalAs"),
+}
+BUGGY_REMOTE_ADDRESSES = [
+    # Bug in JunOS -- info from IPv6 BGP sessions spill over
+    "0.0.0.0",
+    # Bug in earlier Cisco IOS, info from elsewhere (IPv6?) spills over
+    "32.1.7.0",
+]
+
+
+@dataclass
+class BaseBgpRow:
+    peer_state: BgpOperState
+    peer_admin_status: BgpAdminStatus
+    peer_remote_address: IPAddress
+    peer_remote_as: int
+    peer_fsm_established_time: int
+
+
+class BgpStateMonitorTask(Task):
+    """Fetches and stores state information about external BGP sessions."""
+
+    async def run(self):
+        bgp_style = await self._get_bgp_style()
+        if bgp_style != self.device_state.bgp_style:
+            _logger.debug(
+                f"Router {self.device_state.name} changed its BGP style from '{self.device_state.bgp_style}' to "
+                f"'{bgp_style}'"
+            )
+            self.device_state.bgp_style = bgp_style
+        if not bgp_style:
+            return
+
+        local_as = await self._get_local_as(bgp_style=bgp_style)
+        uptime = await self._get_uptime()
+        if not local_as or not uptime:
+            return
+
+        if bgp_style == "juniper":
+            bgp_info = await self._get_juniper_bgp_info()
+        elif bgp_style == "cisco":
+            bgp_info = await self._get_cisco_bgp_info()
+        elif bgp_style == "general":
+            bgp_info = await self._get_general_bgp_info()
```

Comment:
There is enough repetition here that it is tempting to make a class per type (cisco/juniper/generic) and put the classes in a dict, so we get 

```
bgp_info = BGP_TYPES[bgp_style]._get_bgp_info()
```

Probably better to wait with that though, at least until we know everything works correctly.